File: day3_caboose/caboose.go
package main

import (
	"fmt"
	"math"
)

func isPrime(n int) bool {
	if n <= 1 {
		return false
	}
	sqrtN := int(math.Sqrt(float64(n)))
	for i := 2; i <= sqrtN; i++ {
		if n%i == 0 {
			return false
		}
	}
	return true
}

func isCaboose(n int) bool {
	for i := 1; i <= n; i++ {
		x := i*i - i + 41
		if !isPrime(x) {
			return false
		}
	}
	return true
}

func caboosePrimeRatio(n int) float64 {
	primeCount := 0
	for i := 1; i <= n; i++ {
		x := i*i - i + 41
		if isPrime(x) {
			primeCount++
		}
	}
	return float64(primeCount) / float64(n)
}

func main() {
	n := 100
	for i := 1; i < n; i++ {
		if isCaboose(i) {
			fmt.Printf("%d is Caboose\n", i)
		} else {
			fmt.Printf("%d is not Caboose\n", i)
		}
		fmt.Printf("Caboose ratio for %d is %.2f\n", i, caboosePrimeRatio(i))
	}
}

File: day3_caboose/caboose.rs
use std::f64;

fn is_prime(n: i32) -> bool {
    if n <= 1 {
        return false;
    }
    let sqrt_n = (n as f64).sqrt() as i32;
    for i in 2..=sqrt_n {
        if n % i == 0 {
            return false;
        }
    }
    true
}

fn is_caboose(n: i32) -> bool {
    for i in 1..=n {
        let x = i * i - i + 41;
        if !is_prime(x) {
            return false;
        }
    }
    true
}

fn caboose_prime_ratio(n: i32) -> f64 {
    let mut prime_count = 0;
    for i in 1..=n {
        let x = i * i - i + 41;
        if is_prime(x) {
            prime_count += 1;
        }
    }
    prime_count as f64 / n as f64
}

fn main() {
    let n = 100;
    for i in 1..n {
        if is_caboose(i) {
            println!("{} is Caboose", i);
        } else {
            println!("{} is not Caboose", i);
        }
        println!("Caboose ratio for {} is {:.2}", i, caboose_prime_ratio(i));
    }
}
